builder/job.go

```diff
@@ -121,24 +124,52 @@ func Build(d *daemon.Daemon, buildConfig *Config) error {
 	} else if urlutil.IsURL(buildConfig.RemoteURL) {
 		f, err := httputils.Download(buildConfig.RemoteURL)
 		if err != nil {
-			return err
+			return fmt.Errorf("Error downloading remote context %s: %s\n", buildConfig.RemoteURL, err)
 		}
 		defer f.Body.Close()
-		dockerFile, err := ioutil.ReadAll(f.Body)
-		if err != nil {
-			return err
+
+		plen := int(f.ContentLength)
+		if plen > maxPreambleLength {
+			plen = maxPreambleLength
+		}
+		// Some web servers will use application/octet-stream as the default
+		// content type for files without an extension (e.g. 'Dockerfile')
+		// so if we receive this value we better check for text content
+		ct := f.Header.Get("Content-Type")
+		preambleR := bufio.NewReaderSize(f.Body, plen)
+		if len(ct) == 0 || ct == httputils.MimeTypes.OctetStream {
+			ct, _, err = httputils.DetectContentType(preambleR, plen)
+			if err != nil {
+				return fmt.Errorf("Error detecting content type for remote %s: %s\n", buildConfig.RemoteURL, err)
+			}
+		}
+		contentType := selectAcceptableMIME(ct)
+		if len(contentType) == 0 {
+			return fmt.Errorf("builder: Unacceptable content-type %s at remote context %s", ct, buildConfig.RemoteURL)
 		}
 
-		// When we're downloading just a Dockerfile put it in
-		// the default name - don't allow the client to move/specify it
-		buildConfig.DockerfileName = api.DefaultDockerfileName
+		if contentType == httputils.MimeTypes.TextPlain {
+			dockerFile, err := ioutil.ReadAll(f.Body)
+			if err != nil {
+				return err
+			}
 
-		c, err := archive.Generate(buildConfig.DockerfileName, string(dockerFile))
-		if err != nil {
-			return err
+			// When we're downloading just a Dockerfile put it in
+			// the default name - don't allow the client to move/specify it
+			buildConfig.DockerfileName = api.DefaultDockerfileName
+
+			c, err := archive.Generate(buildConfig.DockerfileName, string(dockerFile))
+			if err != nil {
+				return err
+			}
+			context = c
+		} else {
+			// Pass through - this is a pre-packaged context, presumably
+			// with a Dockerfile with the right name inside it.
+			context = ioutil.NopCloser(io.MultiReader(preambleR, f.Body))
```

Comment:
related to previous comments, unless I missed it, I don't see a testcase where the http message is missing a content-type but the body is a Dockerfile.  I'd like to make sure we test for that because that'll hit the code above where we need to inspect the body and then we determine its just plain text and not a tar.